test/fuse_reduce.cpp

```diff
@@ -327,4 +327,158 @@ TEST_CASE(reduce_reduce_broadcast)
+        auto add = add_reduce(
+            p2,
+            "main:reduce_sum0_reshape:main:pointwise0",
+            {xr, yr},
+            {2, 3, 4},
+            [&](auto* rm, const auto& inputs, const auto& axes) {
+                auto rsum  = rm->add_instruction(migraphx::make_op("reduce_sum", {{"axes", axes}}),
+                                                inputs[0]);
+                auto rsumb = rm->add_instruction(
+                    migraphx::make_op("multibroadcast", {{"out_lens", s3.lens()}}), rsum);
+                return add_pointwise(
+                    p2, rm, "main:pointwise0", {rsumb, inputs[1]}, single_pointwise("add"));
+            });
+        auto addr = mm->add_instruction(migraphx::make_op("reshape", {{"dims", s2.lens()}}), add);
+        mm->add_return({addr});
+    }
+    EXPECT(p1.sort() == p2.sort());
+}
+
+TEST_CASE(reduce_reshape_reduce)
+{
+    migraphx::shape s1{migraphx::shape::float_type, {2, 32, 4096}};
+    migraphx::shape s1r{migraphx::shape::float_type, {2, 32, 1}};
+    migraphx::shape s2{migraphx::shape::float_type, {4, 16, 64, 64}};
+    migraphx::shape s2r{migraphx::shape::float_type, {4, 16, 1, 1}};
+    migraphx::shape s3{migraphx::shape::float_type, {2, 2, 16, 64, 64}};
+    migraphx::shape s3r{migraphx::shape::float_type, {2, 2, 16, 1, 1}};
+    migraphx::program p1;
+    {
+        auto* mm       = p1.get_main_module();
+        auto x1        = mm->add_parameter("x1", s1);
+        auto x2        = mm->add_parameter("x2", s1r);
+        auto y         = mm->add_parameter("y", s2);
+        auto rsum1     = mm->add_instruction(migraphx::make_op("reduce_sum", {{"axes", {2}}}), x1);
+        auto rsum1_add = add_pointwise(p1, "main:pointwise0", {rsum1, x2}, single_pointwise("add"));
+
+        auto rsum1_addb = mm->add_instruction(
+            migraphx::make_op("multibroadcast", {{"out_lens", s1.lens()}}), rsum1_add);
+        auto rsum1_sub =
+            add_pointwise(p1, "main:pointwise1", {rsum1_addb, x1}, single_pointwise("sub"));
+        auto rsum1_subr =
+            mm->add_instruction(migraphx::make_op("reshape", {{"dims", s2.lens()}}), rsum1_sub);
+        auto rsum2 =
+            mm->add_instruction(migraphx::make_op("reduce_sum", {{"axes", {2, 3}}}), rsum1_subr);
+        auto rsum2b = mm->add_instruction(
+            migraphx::make_op("multibroadcast", {{"out_lens", s2.lens()}}), rsum2);
+        auto rsum2_add = add_pointwise(p1, "main:pointwise2", {rsum2b, y}, single_pointwise("add"));
+        mm->add_return({rsum2_add});
+    }
+    run_pass(p1);
+    migraphx::program p2;
+    {
+        auto* mm = p2.get_main_module();
+        auto x1  = mm->add_parameter("x1", s1);
+        auto x2  = mm->add_parameter("x2", s1r);
+        auto y   = mm->add_parameter("y", s2);
+        auto x1r = mm->add_instruction(migraphx::make_op("reshape", {{"dims", s3.lens()}}), x1);
+        auto x2r = mm->add_instruction(migraphx::make_op("reshape", {{"dims", s3r.lens()}}), x2);
```

Comment:
I don't understand why it would require s3 of shape {2, 2, 16, 64, 64}. To me it seems like it can just be {4, 16, 64, 64} with reduce axes {2, 3}.

Reply:
This uses the common_shape to compute these shapes which avoids collapsing dimensions. 

The first shape is `{2, 32, 4096}`. If the axes was 1 (instead of 2) then using `{2, 2, 16, 64, 64}` we can change the axes to `{1, 2}`, however if we used `{4, 16, 64, 64}` it would be impossible to set the axis for reduction.

Furthermore, even if we are not reducing on axes 1, one of the inputs could be broadcasting on this axis(ie the shape is `{2, 32, 4096}, {0, 1, 0}`) as well which will lead to extra contiguous being inserted.